Release v1.beta2

mark-a-potts opened the pull request:
This PR includes updates to DA_update and is mostly designed to make the system more portable.

Currently, the submit_cycle.sh script is set to use settings_cycle_test and is expecting input data to be in the "inputs" directory that is in the parent directory of land-offline_workflow. Ultimately, I would like to separate the two, but this is an attempt to get everyone working from a common point and get testing underway.

This has been tested with 2016 data and can be run as follows--

# create a test directory that will be LANDDAROOT
cd land-test
# clone the feature branch recursively
git clone -b feature/release-v1.beta2 --recursive https://github.com/NOAA-EPIC/land-offline_workflow.git
# download the data
wget https://epic-sandbox-srw.s3.amazonaws.com/landda-data-2016.tar.gz
# untar the data
tar xvfz landda-data-2016.tar.gz
# cd into the workflow directory
cd land-offline_workflow/
# load the module files
module use $PWD/modulefiles
module load landda_orion.intel (change to hera for Hera)
# create a build directory and cd into it
mkdir build
cd build
# run the ecbuild command to configure
ecbuild ..
# build everything
make -j 8
#back up to the main directory
cd ..
# submit a sample job that will run two days
sbatch submit_sample_DA_cycle_test.sh
# check on the status
squeue -u $USER
# watch the output go by
tail -f log* err*
# need to hit control-c to exit tail, but then you can check for the background and analysis files in the cycle_land directory
ls -l ../cycle_land/DA_GHCN_test/mem000/restarts/vector/

TO DO:

Set up a test to verify that the current version is working correctly and that future changes don't break anything.
Separate out Inputs from LANDDAROOT variable.
Test on 2020/2021 data.
Organize environment variables so that they are all set with reasonable defaults in either DA sections or land driver sections and can be overridden
Add a "check_environment.sh" script that will check to see if everything needed to run is in place.

mark-a-potts committed:
Sort of working version when using submit_cycle.sh

mark-a-potts committed:
Updates for more portable operations

mark-a-potts committed:
tweaks to run on orion

mark-a-potts committed:
changed settings in submission script

mark-a-potts committed:
added hera modules

ulmononian commented:
following this verbatim, failure on orion after this call:

`/work2/noaa/epic-ps/cbook/land-test/land-offline_workflow/DA_update//do_landDA_release.sh /work2/noaa/epic-ps/cbook/land-test/land-offline_workflow/settings_DA_test`. point of failure is here:

`+ mkdir /work2/noaa/epic-ps/cbook/land-test/outputs/workdir/jedi
mkdir: cannot create directory ‘/work2/noaa/epic-ps/cbook/land-test/outputs/workdir/jedi’: No such file or directory`

err log: /work2/noaa/epic-ps/cbook/land-test/land-offline_workflow/log_noahmp.8925530.log


To get past this issue, line 80 of  https://github.com/NOAA-EPIC/land-DA_update/blob/feature/release-v1.beta2/do_landDA_release.sh can be changed to `  mkdir -p $JEDIWORKDIR`

ulmononian commented:
suggested alteration above gets past that issue, but now hitting:

`+ cp /work2/noaa/epic-ps/cbook/land-test/outputs/workdir/jedi/20160101.180000.coupler.res /work2/noaa/epic-ps/cbook/land-test/outputs/workdir/jedi/mem_neg/20160101.180000.coupler.res
+ echo 'do_landDA: calling create ensemble'
+ python /work2/noaa/epic-ps/cbook/land-test/land-offline_workflow/DA_update//letkf_create_ens.py 20160101.180000 snwdph 30
Traceback (most recent call last):
  File "/work2/noaa/epic-ps/cbook/land-test/land-offline_workflow/DA_update//letkf_create_ens.py", line 1, in <module>
    import numpy as np
ModuleNotFoundError: No module named 'numpy'
+ [[ 1 != 0 ]]
+ echo 'letkf create failed'
+ exit 10
+ [[ 10 != 0 ]]
+ echo 'land DA script failed'
+ exit
~`

modules loaded are from orion.modules, namely:

![Screen Shot 2023-02-09 at 10 00 53 AM](https://user-images.githubusercontent.com/43379611/217898834-bbc41841-1e5f-4d36-8791-fbca01d60189.png)


stack-python/4.12.0 actually points to /apps/miniconda-4.12.0/bin/python, which does not have numpy. since we can't modify a core python install, why are we not pointing to the epic miniconda with `module use /work/noaa/epic-ps/role-epic-ps/miniconda3/modulefiles
module load miniconda3/4.12.0` (which actually also does not have numpy but we can install it there). if it is problematic for us to use the spack stack env built using  /apps/miniconda-4.12.0/bin/python but try to run with EPIC's miniconda, we can re-install the stack, but we might be fine without doing so.

ulmononian commented:
i also had to install netcdf4 in the role.epic miniconda3 (see you have this as a to-do) and adjusted the orion.module file. with the fix to line 80 of `do_landDA_release.sh` and these python updates, the restarts were succesfully generated in `/work2/noaa/epic-ps/cbook/land-test/cycle_land/DA_GHCN_test/mem000/restarts/vector`.

i still think it is confusing that there exists both `workdir` and `outputs` in $LANDDAROOT, primarily because `outputs` then contains another folder called `workdir` which is empty (at least from my run).

the whole test i did is here: `/work2/noaa/epic-ps/cbook/land-test`

ClaraDraper-NOAA commented:
@mark-a-potts I am dealing with an urgent issue for our renalysis replay, and am unlikely to get to this before Monday.

mark-a-potts committed:
Updates for running with singularity. And check_test.sh now works on orion

mark-a-potts committed:
Added test to automatically install netCDF4

barlage commented:
I'm attempting concurrent tests on orion and hera.

hera: there is no hera.modules

orion: it seems there is a permission issue for me when loading modules

```
[orion:/work/noaa/stmp/mbarlage/epic/land-test/land-offline_workflow]$ source orion.modules 
Lmod has detected the following error:  Unable to load module because of error when
evaluating modulefile:

/work/noaa/epic-ps/role-epic-ps/spack-stack/envs/landda-release-1.0-intel/install/modulefiles/intel/2022.0.2/stack-python/3.9.12.lua:
Empty or non-existant file

```
```
[orion:/work/noaa/stmp/mbarlage/epic/land-test/land-offline_workflow]$ more /work/noaa/epic-ps/role-epic-ps/spack-stack/envs/landda-release-1.0-intel/install/modulefiles/intel/2022.0.2/stack-python/3.9.12.lua
/work/noaa/epic-ps/role-epic-ps/spack-stack/envs/landda-release-1.0-intel/install/modulefiles/intel/2022.0.2/stack-python/3.9.12.lua: Permission denied

```

barlage commented on orion.modules:
```diff
@@ -0,0 +1,7 @@
+export JEDI_INSTALL=/work/noaa/epic-ps/role-epic-ps/contrib
+EPICHOME=/work/noaa/epic-ps/role-epic-ps
```
add export to EPICHOME assignment

mark-a-potts replied:
Done.

mark-a-potts committed:
added hera.modules

mark-a-potts committed:
exporting EPICHOME

mark-a-potts committed:
Merge branch 'feature/release-v1.beta2' of https://github.com/NOAA-EPIC/land-offline_workflow into feature/release-v1.beta2

barlage commented on hera.modules:
```diff
@@ -0,0 +1,7 @@
+EPICHOME=/scratch1/NCEPDEV/nems/role.epic
```
add export here too

mark-a-potts commented:
I added in the hera.modules (forgot to add it to my commit earlier) and I chmod'ed all the modules under the role account on Orion, so you should be able to use the orion modules now. Note that they are not the final version for the modules and you need to "source orion.modules" to load them for now. We'll convert them to lua soon.

mark-a-potts committed:
Update hera.modules

added in export for EPICHOME

barlage commented:
I'm getting netcdf4 errors on both hera and orion, I assume similar to @ulmononian , e.g.:
```
++ python -c 'import netCDF4'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'netCDF4'
+ [[ 1 != 0 ]]
+ echo 'no netcdf4, trying to install'

```

Re-ran both again from fresh clones, so logs have changed (same results).

hera: /scratch2/NCEPDEV/land/Michael.Barlage/epic/land-test/land-offline_workflow/err_noahmp.41870402.err
orion: /work/noaa/stmp/mbarlage/epic/land-test/land-offline_workflow/err_noahmp.8942940.err

mark-a-potts commented:
> 

Have you pulled the latest pushes to the repo? There should be a stanza at the beginning of submit_cycle.sh that at least tries to install netCDF4, though maybe it doesn't work from a compute node...

The stanza looks like this--

`python -c 'import netCDF4'`
if [[ $? != 0 ]]; then
  echo 'no netcdf4, trying to install'
  python -m pip install netCDF4
  `python -c 'import netCDF4'`
  if [[ $? != 0 ]]; then
    echo "could not install netCDF4 automatically. Please add netCDF4 module manually and re-run"
    exit
  fi
fi

barlage commented:
I didn't add the full error log but you can see that it tries but fails (the hera run eventually times out I think). 

```
+ python -m pip install netCDF4
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x2b60630d2130>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /simple/netcdf4/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x2b60630d2d30>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /simple/netcdf4/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x2b60630d2ee0>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /simple/netcdf4/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x2b60630f60d0>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /simple/netcdf4/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x2b60630f6280>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /simple/netcdf4/
ERROR: Could not find a version that satisfies the requirement netCDF4 (from versions: none)
ERROR: No matching distribution found for netCDF4

```

mark-a-potts commented:
Okay. I'll have to re-think that. It should be part of the python package that gets loaded to begin with, really.

ulmononian commented:
> Okay. I'll have to re-think that. It should be part of the python package that gets loaded to begin with, really.

@mark-a-potts i will say this again here (i mentioned this in a previous comment): why not just load the miniconda from the role.epic and ensure that netcdf4 is installed there? (i already did this on orion and resolved the issue there; it only stands to be done on hera). since your module file loads "stack-python", it is loading `` on orion but already loads `/scratch1/NCEPDEV/nems/role.epic/miniconda3/4.12.0/bin/python` on hera since that stack was built referencing EPIC's miniconda. 

not sure if a good idea to be installing netcdf into the python that a user is trying to use, assuming they in theory have modified the modules loaded. further, i don't believe we can modify the stack-python currently loaded on orion (/apps/miniconda-4.12.0/bin/python) so a pip install would not work there either.

ulmononian commented:
hera python updated to include netcdf4. @barlage please give it a shot now :) you can always check before you run the job w/ ` python -c "import netCDF4"` but it did import for me.

barlage commented:
@mark-a-potts @ulmononian I believe that I just successfully ran on hera. I needed to make one change in `do_landDA_release.sh`, adding `-p` to line 75:

`mkdir -p $JEDIWORKDIR`

otherwise I think everything was out-of-the-box. When this mkdir change gets in, I will have someone else locally test on hera.

ulmononian commented:
> @mark-a-potts @ulmononian I believe that I just successfully ran on hera. I needed to make one change in `do_landDA_release.sh`, adding `-p` to line 75:
> 
> `mkdir -p $JEDIWORKDIR`
> 
> otherwise I think everything was out-of-the-box. When this mkdir change gets in, I will have someone else locally test on hera.

@barlage i had the same issue in my testing yesterday on orion (see my first comment here) but could not suggest the change in my review since the script is part of DA_update. had the same issue when i built/ran with the newest commits today. had to add the `-p` when creating the `$JEDIWORKDIR`.

ulmononian commented:
linking issues are back (see log here /work2/noaa/epic-ps/cbook/land-test-2/land-offline_workflow/log_noahmp.8944368.log). i noticed `find_package( fv3-bundle REQUIRED)` has been commented out of the land-offline_workflow CMakeLists.txt, so guessing that is why it can't find `${fv3-bundle_BASE_DIR}` in line 52 (https://github.com/NOAA-EPIC/land-offline_workflow/blob/f0141f35469d523d6f86b61b7623affb0c63f9dc/CMakeLists.txt#L53). 

@mark-a-potts i know you mentioned we may have to cp the files rather than softlink at this point, but i will just throw my solution out there one more time: https://github.com/NOAA-EPIC/land-DA_update/pull/1. just run the links upon the first cycle run and does not link again if they exist.

mark-a-potts commented:
Ah yes, I forgot I changed that. You need to run ecbuild with this 
command (after setting EPICHOME correctly for your platform) now--

ecbuild -Dfv3-bundle_BASE_DIR=$EPICHOME/contrib/fv3-bundle ..

We should probably switch this around to use an environment variable 
that is set in the modulefile for the platform and gets picked up by CMake.

-M


On 2/10/23 12:40 PM, Cameron Book wrote:
>
> linking issues are back (see log here 
> /work2/noaa/epic-ps/cbook/land-test-2/land-offline_workflow/log_noahmp.8944368.log). 
> i noticed |find_package( fv3-bundle REQUIRED)| has been commented out 
> of the land-offline_workflow CMakeLists.txt, so guessing that is why 
> it can't find |${fv3-bundle_BASE_DIR}| in line 52 
> (https://github.com/NOAA-EPIC/land-offline_workflow/blob/f0141f35469d523d6f86b61b7623affb0c63f9dc/CMakeLists.txt#L53).
>
> @mark-a-potts <https://github.com/mark-a-potts> i know you mentioned 
> we may have to cp the files rather than softlink at this point, but i 
> will just throw my solution out there one more time: 
> NOAA-EPIC/land-DA_update#1 
> <https://github.com/NOAA-EPIC/land-DA_update/pull/1>. just run the 
> links upon the first cycle run and does not link again if they exist.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/NOAA-PSL/land-offline_workflow/pull/25#issuecomment-1426132246>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AH4Q2UXGPXZXNCN46WPJHJTWWZ4PPANCNFSM6AAAAAAUW2EMLE>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>
-- 
--
Mark A. Potts, Ph.D.
NOAA EPIC Lead Software Engineer
RedLine Performance Solutions, LLC
Phone 202-744-9469
***@***.***
***@***.***

mark-a-potts committed:
updated do_landDA_release

mark-a-potts commented:
I just pushed a change to DA_update that changes that "python" to 
"${PYTHON}" which should work (hopefully). The change also added in the 
"-p" to the mkdir calls.

-M

On 2/9/23 1:01 PM, Cameron Book wrote:
>
> suggested alteration above gets past that issue, but now hitting:
>
> `+ cp 
> /work2/noaa/epic-ps/cbook/land-test/outputs/workdir/jedi/20160101.180000.coupler.res 
> /work2/noaa/epic-ps/cbook/land-test/outputs/workdir/jedi/mem_neg/20160101.180000.coupler.res
>
>   * echo 'do_landDA: calling create ensemble'
>   * python
>     /work2/noaa/epic-ps/cbook/land-test/land-offline_workflow/DA_update//letkf_create_ens.py
>     20160101.180000 snwdph 30
>     Traceback (most recent call last):
>     File
>     "/work2/noaa/epic-ps/cbook/land-test/land-offline_workflow/DA_update//letkf_create_ens.py",
>     line 1, in
>     import numpy as np
>     ModuleNotFoundError: No module named 'numpy'
>   * [[ 1 != 0 ]]
>   * echo 'letkf create failed'
>   * exit 10
>   * [[ 10 != 0 ]]
>   * echo 'land DA script failed'
>   * exit
>     ~`
>
> modules loaded are from orion.modules, namely:
>
> Screen Shot 2023-02-09 at 10 00 53 AM 
> <https://user-images.githubusercontent.com/43379611/217898834-bbc41841-1e5f-4d36-8791-fbca01d60189.png>
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/NOAA-PSL/land-offline_workflow/pull/25#issuecomment-1424598647>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AH4Q2USV3FNFFPNSXDSLDCLWWUWGTANCNFSM6AAAAAAUW2EMLE>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>
-- 
--
Mark A. Potts, Ph.D.
NOAA EPIC Lead Software Engineer
RedLine Performance Solutions, LLC
Phone 202-744-9469
***@***.***
***@***.***

barlage commented:
@ulmononian I did see your earlier comment and assumed it was the same issue, but had lost my orion session and was too lazy to log back in to confirm that your line 80 was my line 75.

ulmononian commented:
> @ulmononian I did see your earlier comment and assumed it was the same issue, but had lost my orion session and was too lazy to log back in to confirm that your line 80 was my line 75.

lol. just glad you were able to reproduce the error to confirm not the usual user error on my side. @mark-a-potts thanks for pushing those changes!!

mark-a-potts committed:
change module files to lua and put them in modulefiles

mark-a-potts committed:
Added check for JEDI_INSTALL env variable which is used with make_links.sh
Also updated check_test.sh to find comparison data under EPICHOME on
Hera and Orion.

ClaraDraper-NOAA commented on check_test.sh:
```diff
@@ -1,18 +1,20 @@
+source ./settings_cycle_test
 
-TEST_BASEDIR=/scratch2/BMC/gsienkf/Clara.Draper/DA_test_cases/land-offline_workflow/DA_IMS_test/output/mem000/restarts/vector/
+TEST_BASEDIR=/work/noaa/epic-ps/role-epic-ps/landda/cycle_land/DA_GHCN_test/mem000/restarts/vector
```
If we want to use different values here, I suggest that we put the BASEDIR in the settings file.

mark-a-potts replied:
I updated this to use ${EPICHOME} as the base path unless TEST_BASEDIR is defined. We can do this with most/all of the variables I have been setting in these scripts I think.

ClaraDraper-NOAA commented on do_submit_cycle.sh:
```diff
@@ -96,18 +96,37 @@ ln -sf ${MEM_MODL_OUTDIR}/noahmp ${MEM_WORKDIR}/noahmp_output
+#       exit 10 
+#    fi 
+#fi 
+    # if restart not in experiment out directory, copy the restarts from the ICSDIR
```
I intentionally removed this code (which is designed to populate an ensemble of restarts from a single restart) as it's not relevant here. Please remove again.

mark-a-potts replied:
I don't think the job will run without it when you are starting from a fresh set of inputs.  My intention was to be able to start an experiment with no cycle_land or workdir in place yet. Is there a better way to find a restart file?

ClaraDraper-NOAA commented on submit_cycle.sh:
```diff
@@ -1,22 +1,68 @@
+set -x
+export LANDDAROOT=${LANDDAROOT:-`dirname $PWD`}
+
+`python -c 'import netCDF4'`
```
Do we really need to check this every time this job submits? Wouldn't this be better in a set-up script (or even just as an assumed software requirement).

mark-a-potts replied:
Yeah, I will remove that. We updated the python modules so that there is no need to check now.

ClaraDraper-NOAA commented on submit_cycle.sh:
```diff
@@ -1,22 +1,68 @@
+  export JEDI_EXECDIR=${LANDDAROOT}/land-offline_workflow/singularity/bin
+  #we need to have intelmpi loaded on the host system to run the workflow. Try to load it here.
+  #TODO--figure out a way to make sure we have intelmpi loaded or don't let the workflow start
+  module try-load impi
```
Likewise - can we load these modules once?  At the very least, can we move all of the module loading to a separate file?

mark-a-potts replied:
That is only for running with a container. It may make sense to separate the submission script for containers from the one used on bare metal.

ClaraDraper-NOAA commented on submit_cycle.sh:
```diff
@@ -1,22 +1,68 @@
+fi
+export BUILDDIR=${BUILDDIR:-${LANDDAROOT}/land-offline_workflow/build}
+export CYCLEDIR=$(pwd) 
+export PYTHON=`which python3`
```
Should we be over-writing the value of PYTHON set above for singularity?

mark-a-potts replied:
That is a good question. I think we should not be. I'll change that.

ClaraDraper-NOAA commented on submit_cycle.sh:
```diff
@@ -1,22 +1,68 @@
 THISDATE=$STARTDATE
 date_count=0
 
+vec2tileexec=${BUILDDIR}/bin/vector2tile_converter.exe
```
The executables are all specified in do_submitcycle.sh. Why do they need to be repeated here?

mark-a-potts replied:
Perhaps I misunderstood which file to start with. I am not using do_submit_cycle.sh at all. The way things are set up right now, there is only one test/experiment to run and it is started using "sbatch submit_cycle.sh". That script is hard coded to source settings_cycle_test and runs do_landda_release.sh in the DA_update directory.  

We can definitely re-factor to change the functionality around, though.

mark-a-potts replied:
I did rename everything to the following--
submit_sample_DA_cycle_test.sh 
settings_sample_DA_cycle_test 
DA_update/do_sample_landDA_test.sh

ClaraDraper-NOAA replied:
The code is designed to be submitted using the do_submit_cycle.sh script. At the moment, this PR is not using that script, and has copied much of this functionality into submit_cycle.sh. Can you please switch back to using do_submit_cycle.sh. Otherwise, it's going to be very difficult to maintain two different versions.

mark-a-potts replied:
What is the functionality that you are looking to preserve? We have moved to using the submit_sample_DA_cycle_test.sh as the sample submission script. I am unclear on the difference between the original submit_cycle.sh vs do_submit_cycle.sh.

ClaraDraper-NOAA replied:
The intention is that jobs are submitted using do_submit_cycle.sh, which stages the necessary files, sets any variables needed (basically does everything that needs to be done once per experiment). It then calls submit_cycle.sh, which cycles through the forecasts and DA, and will re-submit itself if need be (i.e., if we want to submit a number of smaller jobs, rather than one big job). At the moment, the PR has changes to do_submit_cycle.sh and submit_cycle.sh, as well as the addition of new submission scripts. Are you still using do_submit_cycle.sh and submit_cycle.sh, or is your intention that the new scripts replace those? we shouldn't really need separate submit scripts for different jobs - I only created the test one for convenience.

The changes to the scripts are much more extensive than I was expecting ( in a separate project, we have altered my scripts to run on the cloud using containers, by creating new settings files, changing the executable calls, and making minimal other changes, as needed). I am almost certain that the initial functionality of do_submit_cycle.sh  and submit_cycle.sh on hera has not been retained. It is going to be very difficult / a lot of work for me to merge these changes into my develop branch. Either we need to revert this PR to follow the original design of the scripts, or I suggest I just merge it without really looking at it with the understanding that this work will not make it back into develop (unless it's fixed up later).

ClaraDraper-NOAA commented on submit_cycle.sh:
```diff
@@ -1,22 +1,68 @@
+export BUILDDIR=${BUILDDIR:-${LANDDAROOT}/land-offline_workflow/build}
+export CYCLEDIR=$(pwd) 
+export PYTHON=`which python3`
+export OBSDIR=${LANDDAROOT}/inputs/DA
```
OBSDIR relates to the DA. It would be better to leave it with the DA code (settings_DA or do_landDA.sh)

mark-a-potts replied:
So, is the intention for settings_DA_* to be using DA while the settings files without "DA" in them to simply run a forecast with no DA? Maybe it makes sense for me to rename the scripts I am using for this test case to something like "submit_sample_DA_cycle_test.sh" and "settings_sample_DA_cycle_test" and finally "do_sample_landDA_test.sh"?

ClaraDraper-NOAA commented on submit_cycle.sh:
```diff
@@ -49,7 +95,12 @@ while [ $date_count -lt $cycles_per_job ]; do
-    
+   
+    # compute the restart frequency, run_days and run_hours
+    FREQ=$(( 3600 * $FCSTHR ))
```
Also done in do_submit_cycle.sh. Why are we doing it again here?

ClaraDraper-NOAA replied:
Same comment for code inserted below.

ClaraDraper-NOAA commented on submit_cycle_release.sh:
```diff
@@ -0,0 +1,413 @@
+
+# copy initial conditions
+n_ens=1
+while [ $n_ens -le $ensemble_size ]; do
```
There should be no reference to n_ens or ensemble_size anymore

mark-a-potts replied:
That script was a leftover from v1.beta and I had been using it for reference. I will remove it.

ClaraDraper-NOAA reviewed (changes requested):
Thanks for this Mark.

It's hard to review the code without knowing how you're calling it - it looks like a lot of code has moved form do_submit_cycle to submit_cycle. 

Are you using submit_cycle, do_submit_cycle, or submit_release? Could you please delete anything that is not being used?  Likewise, for the settings file - include only the one you use for your test, and a template.

Also, there's some old ensemble-code still in there - it looks like it hasn't been fully re-based to my EPIC release branch.

mark-a-potts committed:
Addressing some of the issues raised by Clara

mark-a-potts committed:
Updated submodule pointer and removed old module files

mark-a-potts committed:
updated check_test to allow parameter for settings file

mark-a-potts committed:
updated size of strings in ufs-land-driver to facilitate testing

mark-a-potts committed:
updated size of strings in ufs-land-driver to facilitate testing

mark-a-potts committed:
added return codes for check_test failures for CI/CD

EdwardSnyder-NOAA committed:
Update landdaroot to landda_inputs

EdwardSnyder-NOAA committed:
split inputs and outputs from landdaroot var

EdwardSnyder-NOAA committed:
Update landdaroot to landda_inputs

EdwardSnyder-NOAA committed:
split inputs and outputs from landdaroot var

EdwardSnyder-NOAA committed:
fix submodule

EdwardSnyder-NOAA committed:
update submodule

EdwardSnyder-NOAA committed:
updated submodule

mark-a-potts committed:
Updates to run with 2020-2021 era5 forcing

mark-a-potts committed:
Merge branch 'feature/release-v1.beta2' of https://github.com/NOAA-EPIC/land-offline_workflow into feature/release-v1.beta2

mark-a-potts committed:
Separated submit_sample into do_submit_cycle.sh and submit_cycle.sh

mark-a-potts commented:
This morning, I took the monolithic submit_sample_DA_cycle_test.sh and 
separated it into a submit_cycle.sh and do_submit_cycle.sh that should 
track largely with what you were doing before. I removed the setup of 
the directories from do_submit_cycle.sh since that is now being done in 
submit_cycle.sh and there are a couple of other changes with regards to 
restart files which may need to be changed. If these changes look 
reasonable, it would be really helpful to approve them so that we can 
get the release branch cut and all work from that. There are several 
other PRs that are all waiting to be done and we can make changes to the 
release branch according to what you would like to see there.

-M

On 2/15/23 11:30 AM, ClaraDraper-NOAA wrote:
>
> ***@***.**** commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In submit_cycle.sh 
> <https://github.com/NOAA-PSL/land-offline_workflow/pull/25#discussion_r1107361521>:
>
> >  THISDATE=$STARTDATE
>   date_count=0
>   
> +vec2tileexec=${BUILDDIR}/bin/vector2tile_converter.exe
>
> The intention is that jobs are submitted using do_submit_cycle.sh, 
> which stages the necessary files, sets any variables needed (basically 
> does everything that needs to be done once per experiment). It then 
> calls submit_cycle.sh, which cycles through the forecasts and DA, and 
> will re-submit itself if need be (i.e., if we want to submit a number 
> of smaller jobs, rather than one big job). At the moment, the PR has 
> changes to do_submit_cycle.sh and submit_cycle.sh, as well as the 
> addition of new submission scripts. Are you still using 
> do_submit_cycle.sh and submit_cycle.sh, or is your intention that the 
> new scripts replace those? we shouldn't really need separate submit 
> scripts for different jobs - I only created the test one for convenience.
>
> The changes to the scripts are much more extensive than I was 
> expecting ( in a separate project, we have altered my scripts to run 
> on the cloud using containers, by creating new settings files, 
> changing the executable calls, and making minimal other changes, as 
> needed). I am almost certain that the initial functionality of 
> do_submit_cycle.sh and submit_cycle.sh on hera has not been retained. 
> It is going to be very difficult / a lot of work for me to merge these 
> changes into my develop branch. Either we need to revert this PR to 
> follow the original design of the scripts, or I suggest I just merge 
> it without really looking at it with the understanding that this work 
> will not make it back into develop (unless it's fixed up later).
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/NOAA-PSL/land-offline_workflow/pull/25#discussion_r1107361521>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AH4Q2UVX744VKOV75GILYXDWXUABBANCNFSM6AAAAAAUW2EMLE>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>
-- 
--
Mark A. Potts, Ph.D.
NOAA EPIC Lead Software Engineer
RedLine Performance Solutions, LLC
Phone 202-744-9469
***@***.***
***@***.***

ClaraDraper-NOAA commented:
> This morning, I took the monolithic submit_sample_DA_cycle_test.sh and separated it into a submit_cycle.sh and do_submit_cycle.sh that should track largely with what you were doing before. I removed the setup of the directories from do_submit_cycle.sh since that is now being done in submit_cycle.sh and there are a couple of other changes with regards to restart files which may need to be changed. If these changes look reasonable, it would be really helpful to approve them so that we can get the release branch cut and all work from that. There are several other PRs that are all waiting to be done and we can make changes to the release branch according to what you would like to see there.
> -M
> On 2/15/23 11:30 AM, ClaraDraper-NOAA wrote: ***@***.**** commented on this pull request. ------------------------------------------------------------------------ In submit_cycle.sh <[#25 (comment)](https://github.com/NOAA-PSL/land-offline_workflow/pull/25#discussion_r1107361521)>: > THISDATE=$STARTDATE date_count=0 +vec2tileexec=${BUILDDIR}/bin/vector2tile_converter.exe The intention is that jobs are submitted using do_submit_cycle.sh, which stages the necessary files, sets any variables needed (basically does everything that needs to be done once per experiment). It then calls submit_cycle.sh, which cycles through the forecasts and DA, and will re-submit itself if need be (i.e., if we want to submit a number of smaller jobs, rather than one big job). At the moment, the PR has changes to do_submit_cycle.sh and submit_cycle.sh, as well as the addition of new submission scripts. Are you still using do_submit_cycle.sh and submit_cycle.sh, or is your intention that the new scripts replace those? we shouldn't really need separate submit scripts for different jobs - I only created the test one for convenience. The changes to the scripts are much more extensive than I was expecting ( in a separate project, we have altered my scripts to run on the cloud using containers, by creating new settings files, changing the executable calls, and making minimal other changes, as needed). I am almost certain that the initial functionality of do_submit_cycle.sh and submit_cycle.sh on hera has not been retained. It is going to be very difficult / a lot of work for me to merge these changes into my develop branch. Either we need to revert this PR to follow the original design of the scripts, or I suggest I just merge it without really looking at it with the understanding that this work will not make it back into develop (unless it's fixed up later). — Reply to this email directly, view it on GitHub <[#25 (comment)](https://github.com/NOAA-PSL/land-offline_workflow/pull/25#discussion_r1107361521)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AH4Q2UVX744VKOV75GILYXDWXUABBANCNFSM6AAAAAAUW2EMLE>. You are receiving this because you were mentioned.Message ID: ***@***.***>
> -- -- Mark A. Potts, Ph.D. NOAA EPIC Lead Software Engineer RedLine Performance Solutions, LLC Phone 202-744-9469 ***@***.*** ***@***.***

Thanks Mark.  What's the reason for moving the directory set-up into submit_cycle.sh? Let me know when you've pushed something that you want me to look at.

mark-a-potts committed:
Updating check_test to work with new settings files.

mark-a-potts commented:
@ClaraDraper-NOAA The changes are in place now. The do_submit_cycle.sh script defaults to using settings_DA_cycle_gdas, but will also work with settings_DA_cycle_era5 (using 2020/21 data). Any other settings files may need to be tweaked, but haven't been tested.

ClaraDraper-NOAA commented:
What's the reason for moving the directory set-up into submit_cycle.sh?

mark-a-potts commented:
Subdirectories were being created in submit_cycle.sh as well as top level in do_submit_cycle.sh, so the creation was all consolidated.

ClaraDraper-NOAA commented:
> Subdirectories were being created in submit_cycle.sh as well as top level in do_submit_cycle.sh, so the creation was all consolidated.

In my original submit_cycle.sh there are no directories created. I will do a formal review later today, but unless there's a reason it needs to be moved, I going to ask you to move all of the directory creation back to do_submit_cycle.sh. I'm trying to minimize the code changes, to make it easier to merge.

mark-a-potts committed:
moved directory creation back to do_submit

mark-a-potts committed:
Removed extraneous settings files and module files for clarity.
Might want to add a settings template later.

mark-a-potts committed:
added check for single member start

ClaraDraper-NOAA commented on submit_cycle.sh:
```diff
@@ -68,18 +68,32 @@ while [ $date_count -lt $cycles_per_job ]; do
 
     # copy restarts into work directory
     rst_in=${LANDDA_INPUTS}/restarts/${atmos_forc}/ufs_land_restart.${YYYY}-${MM}-${DD}_${HH}-00-00.nc 
+    rst_in_single=${LANDDA_INPUTS}/single/output/modl/restarts/vector/ufs_land_restart.${YYYY}-${MM}-${DD}_${HH}-00-00.nc 
```
It should be possible to generalise this - keep the old RSTRDIR (or whatever I called it), and set RSTRDIR to LANDDA_INPUTS in your case.

mark-a-potts replied:
The code in your EPIC1.0 branch pointed to an experiment directory output that wouldn't yet exist for a new user. I will change it to look first for the restart in the experiment directory and, if it is not found, look under LANDDA_INPUTS.

ClaraDraper-NOAA commented on submit_cycle.sh:
```diff
@@ -68,18 +68,32 @@ while [ $date_count -lt $cycles_per_job ]; do
-      echo "Missing restart file"
-    else 
-      cp $rst_in $rst_out 
+    # if restart not in experiment out directory, copy the restarts from the ICSDIR
```
This is the old ensemble code, and has been removed in my latest branch. Please remove it / update to my latest code.

mark-a-potts replied:
Done.

ClaraDraper-NOAA commented on check_test.sh:
```diff
@@ -1,36 +1,45 @@
 
 # get OUTDIR
-source settings_cycle_test
+export LANDDAROOT=${LANDDAROOT:-`dirname $PWD`}
```
This script can no longer be used for my test on hera. Please either generalize it, so that it can be used for both, or create a separate script for your test. 

Also, where possible, please move the specification of all variables into the settings file. 

It is also not necessary to allow for previous values VAR=${VAR:-"newvalue"} here, since none of these variables will have been set.

mark-a-potts replied:
The original script doesn't work for me, so I don't know how to generalize. I will make a new script for our test.

ClaraDraper-NOAA commented on do_submit_cycle.sh:
```diff
@@ -1,12 +1,44 @@
 
+set -x
+#Set defaults
+export LANDDAROOT=${LANDDAROOT:-`dirname $PWD`}
```
I'm confused by the directories. There seems to be lots of new directory variables introduced that are replicates of pre-existing directory variables. You should be able to just use my pre-existing names, and set them appropriately.

Is LANDDAROOT the same as the pre-existing CYCLEDIR, or a directory up? Both are set to the current directory, but then LAND_OFFLINE_WORKFLOW is set to $LANDDAROOT/land-offline_workflow (directory where the cycling scripts are - in my scripts, we call do_submit_cycle.sh from this directory, so CYCLEDIR is also the directory where the cycling scripts are).

mark-a-potts replied:
Yes, LANDDAROOT is the parent directory of land-offline_workflow. I'll change the LAND_OFFLINE_WORKFLOW variable to be CYCLEDIR. It was not obvious to me that CYCLEDIR was always going to be the same as the source directory of land-offline_workflow.

ClaraDraper-NOAA commented on do_submit_cycle.sh:
```diff
@@ -1,12 +1,44 @@
+
+#Change some variables if working with a container
+if [[ ${USE_SINGULARITY} =~ yes ]]; then
+  EPICHOME=/opt
```
I think it'd be better to put these into a separate file and source it (similar to my module files).

mark-a-potts replied:
Agreed. We are working on some sanity checks that will get run at the top of the do_submit_release.sh script and will move a lot of this functionality there.

ClaraDraper-NOAA commented on do_submit_cycle.sh:
```diff
@@ -1,12 +1,44 @@
 if [[ $# -gt 0 ]]; then 
     config_file=$1
 else
-    config_file=settings
```
Please change this back. settings is the default, you can specify your input file as an argument to over-ride it.

mark-a-potts replied:
Done.

ClaraDraper-NOAA commented on do_submit_cycle.sh:
```diff
@@ -22,17 +54,17 @@ export KEEPWORKDIR="YES"
 
 ############################
 # load modules 
-
```
Please add this file back, and add an if statement so that this is only called in my case (not cmake?)

ClaraDraper-NOAA commented on do_submit_cycle.sh:
```diff
@@ -22,17 +54,17 @@ export KEEPWORKDIR="YES"
 
-export vec2tileexec=${CYCLEDIR}/vector2tile/vector2tile_converter.exe
-export LSMexec=${CYCLEDIR}/ufs-land-driver/run/ufsLand.exe 
+export vec2tileexec=${BUILDDIR}/bin/vector2tile_converter.exe
```
These locations will only work for CMake. I'm not sure what to do here  - if we're going to switch to CMake as the only option, then we need to test that we can reproduce the older results - in which case, we need to retain both options for now. Maybe it's enough to keep all the files in place, and we can go in and manually change things like paths when we do the testing?

mark-a-potts replied:
We can address this in a couple of different ways. I can set the script to find the executables in either location, or we can add an "install" step to the CMake config that will install the executables where you used to put them. However, I would strongly suggest that CMake/ecbuild should be the only supported build option. For the moment, I'll set the script to find the executables wherever they may be.

mark-a-potts replied:
The do_submit_cycle script can now find the non-cmake built executables, but they won't run because they are built with different versions of libraries. The cmake-based build uses a single consistent set of modules, so they only need to be loaded once.

ClaraDraper-NOAA commented on CMakeLists.txt:
```diff
@@ -42,12 +48,15 @@ set( LANDDA_LINKER_LANGUAGE CXX)
 add_subdirectory( ufs-land-driver )
 add_subdirectory( DA_update/add_jedi_incr )
-add_subdirectory( DA_update/IMS_proc )
+#add_subdirectory( DA_update/IMS_proc )
```
delete the comment.

ClaraDraper-NOAA commented on CMakeLists.txt:
```diff
@@ -31,6 +36,7 @@ find_package( NetCDF REQUIRED COMPONENTS Fortran)
 #find_package( ioda   1.0.0    REQUIRED)
```
delete the comment block.

ClaraDraper-NOAA commented on .gitmodules:
```diff
@@ -1,13 +1,15 @@
 	#url = https://github.com/barlage/ufs-land-driver.git
-        url = https://github.com/ClaraDraper-NOAA/ufs-land-driver.git
-        branch = feature/limit_mfsno
+#       url = https://github.com/ClaraDraper-NOAA/ufs-land-driver.git
```
delete the commented line. And below.

ClaraDraper-NOAA commented on check_test.sh:
```diff
@@ -1,36 +1,45 @@
-source settings_cycle_test
+export LANDDAROOT=${LANDDAROOT:-`dirname $PWD`}
+export SETTINGS_FILE=${1:-"settings_DA_cycle_gdas"}
+export CYCLE_LAND=${CYCLE_LAND:-${LANDDAROOT}/cycle_land}
```
if this is the output directory, can we give it a different name? "cycle_land" is just me temporary output name, and not particularly well chosen (sounds like the cycle directory).

ClaraDraper-NOAA replied:
... I don't see CYCLE_LAND used below. Can it be deleted?

mark-a-potts replied:
I've renamed the variable LANDDA_EXPTS and the actual directory to landda_expts. Can change that to anything else you would prefer.

mark-a-potts replied:
The variable is picked up in the settings file, but could probably be move into that file. I'll check.

ClaraDraper-NOAA commented on check_test.sh:
```diff
@@ -1,36 +1,45 @@
+source ./${SETTINGS_FILE}
 
-TEST_BASEDIR=/scratch2/BMC/gsienkf/Clara.Draper/DA_test_cases/land-offline_workflow/DA_IMS_test/output/mem000/restarts/vector/
+EPICTESTS=${EPICHOME}/landda/cycle_land/DA_GHCN_test/mem000/restarts/vector
```
Unless there's a reason to add a new variable, please remove EPICTESTS, and use TEST_BASEDIR. You can define TEST_BASEDIR in your settings file.

ClaraDraper-NOAA commented on do_submit_cycle.sh:
```diff
@@ -1,12 +1,44 @@
+export LANDDAROOT=${LANDDAROOT:-`dirname $PWD`}
+export LANDDA_INPUTS=${LANDDA_INPUTS:-${LANDDAROOT}/inputs}
+export LAND_OFFLINE_WORKFLOW=${LAND_OFFLINE_WORKFLOW:-${LANDDAROOT}/land-offline_workflow}
+export CYCLE_LAND=${CYCLE_LAND:-${LANDDAROOT}/cycle_land}
```
This is the output directory, right? Let's change the name to indicate that (also, I will already have an output directory defined - please just use that variable).

mark-a-potts replied:
The OUTDIR variable includes the experiment name. The CYCLE_LAND variable (now called LANDDA_EXPTS) is the parent directory of OUTDIR. Personally, I think OUTDIR should be the parent and LANDDA_EXPT or LANDDA_EXPT_DIR should have the experiment name, but I think that would break things you have in place. We can name them whatever you like, though.

ClaraDraper-NOAA commented on do_submit_cycle.sh:
```diff
@@ -1,12 +1,44 @@
+export CYCLE_LAND=${CYCLE_LAND:-${LANDDAROOT}/cycle_land}
+export PYTHON=`which python3`
+export BUILDDIR=${BUILDDIR:-${LAND_OFFLINE_WORKFLOW}/build}
+export PATH=$PATH:./
```
Another variable for the current directory.

ClaraDraper-NOAA commented on do_submit_cycle.sh:
```diff
@@ -1,12 +1,44 @@
+export LANDDA_INPUTS=${LANDDA_INPUTS:-${LANDDAROOT}/inputs}
+export LAND_OFFLINE_WORKFLOW=${LAND_OFFLINE_WORKFLOW:-${LANDDAROOT}/land-offline_workflow}
+export CYCLE_LAND=${CYCLE_LAND:-${LANDDAROOT}/cycle_land}
+export PYTHON=`which python3`
```
These should be defined in the settings file.

ClaraDraper-NOAA commented on do_submit_cycle.sh:
```diff
@@ -93,22 +125,6 @@ if [[ ! -e ${MEM_MODL_OUTDIR}/restarts/ ]]; then  # subdirectories
 fi
 ln -sf ${MEM_MODL_OUTDIR}/noahmp ${MEM_WORKDIR}/noahmp_output 
 
-# copy ICS into restarts, if needed 
```
Please add this back in - the restart files should be copied across from this script.

ClaraDraper-NOAA commented on submit_cycle.sh:
```diff
@@ -1,11 +1,11 @@
-#!/bin/bash -le 
+#!/bin/bash  
 #SBATCH --job-name=offline_noahmp
-#SBATCH --account=gsienkf
```
We should probably add a mechanism for the use to specify the account. Worry about this later though?

mark-a-potts replied:
Yes. Maybe we can add it to one of the sanity checks that get run.

ClaraDraper-NOAA commented on submit_cycle.sh:
```diff
@@ -49,7 +49,12 @@ while [ $date_count -lt $cycles_per_job ]; do
     DP=`echo $PREVDATE | cut -c7-8`
     HP=`echo $PREVDATE | cut -c9-10`
-    
+   
```
still not sure why we're doing this here.

mark-a-potts replied:
Removed.

ClaraDraper-NOAA commented on submit_cycle.sh:
```diff
@@ -155,10 +181,11 @@ while [ $date_count -lt $cycles_per_job ]; do
+    set -x 
     # update model namelist 
-    cp  ${CYCLEDIR}/template.ufs-noahMP.namelist.${atmos_forc}  ufs-land.namelist
+    cp  ${CYCLEDIR}/template.ufs-noahMP.namelist.release.${atmos_forc}  ufs-land.namelist
```
can we add "release" to $atmos_forc?

mark-a-potts replied:
Not really. the ${atmos_forc} variable is part of the directory structure for the forcing. I have renamed the template to remove the "release" part, though.

ClaraDraper-NOAA commented on template.tile2vector:
```diff
@@ -18,7 +18,7 @@
   
 ! Path for static file  
-  static_filename = "/scratch1/NCEPDEV/stmp2/Michael.Barlage/forcing/CXXRES/static/ufs-land_CXXRES_static_fields.nc"
+  static_filename = "LANDDA_INPUTS/forcing/MODELTYPE/static/ufs-land_CXXRES_static_fields.nc"
```
Please create a new FORCING variable, so that the assumption is not built in that everything is in the same directory.

mark-a-potts replied:
Changed this to MODEL_FORCING

ClaraDraper-NOAA reviewed (changes requested):
Thanks Mark. It's reasonably close - my main concern is that there are lots of duplicate directory names introduced, that are not necessary (and will also break the scripts if I try to run the old way on hera). You should be able to just set my current variables appropriately, without introducing (many) new variables.

ulmononian committed:
Change Hera reference to Orion in landda_orion.intel.lua

mark-a-potts committed:
Release branch created with updated submodules. Not ready for release

mark-a-potts committed:
Updates to address code review from Clara

ulmononian committed:
Update module files and introduced module_check.sh script to load modules at run time, check that modules are loaded properly, set compiler wrappers, and JEDI install path.
